File: apps/desktop/src-tauri/src/vector.rs
//! Vector/semantic search stub — mirrors the JS `lib/vector.ts` embedding.
/// Local embedding using hashed TF + synonym groups, no ONNX dependency.
/// This module is the Rust counterpart for `sqlite-vec` hybrid search.
/// A future PR will swap this for `all-MiniLM` via `onnx` + `sqlite-vec`.

use serde::{Deserialize, Serialize};

const DIM: usize = 64;

/// Small synonym groups — must stay in sync with `apps/desktop/src/lib/vector.ts`.
const SYNONYM_GROUPS: &[&[&str]] = &[
    &["vacation", "holiday", "trip", "getaway", "travel"],
    &["car", "automobile", "vehicle", "auto"],
    &["salary", "pay", "wage", "income", "compensation"],
    &["hello", "hi", "hey", "greetings"],
    &["react", "frontend", "ui"],
    &["rust", "cargo", "tauri"],
];

fn hash_token(token: &str) -> u32 {
    let mut h: u32 = 2166136261;
    for b in token.bytes() {
        h ^= b as u32;
        h = h.wrapping_mul(16777619);
    }
    h
}

fn tokenize(text: &str) -> Vec<String> {
    text.to_lowercase()
        .split(|c: char| !c.is_ascii_alphanumeric())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

/// Deterministic embedding — 64 dims, L2-normalized, synonym-aware.
pub fn embed(text: &str) -> Vec<f32> {
    let mut vec = vec![0f32; DIM];
    let tokens = tokenize(text);
    if tokens.is_empty() {
        return vec;
    }
    for token in tokens {
        let bucket = (hash_token(&token) as usize) % DIM;
        vec[bucket] += 1.0;
        // synonym group boost
        for (gid, group) in SYNONYM_GROUPS.iter().enumerate() {
            if group.contains(&token.as_str()) {
                let group_bucket = (hash_token(&format!("__group_{}", gid)) as usize) % DIM;
                vec[group_bucket] += 1.5;
                let canon = group[0];
                let canon_bucket = (hash_token(canon) as usize) % DIM;
                if canon_bucket != bucket {
                    vec[canon_bucket] += 0.7;
                }
                break;
            }
        }
    }
    let norm: f32 = vec.iter().map(|v| v * v).sum::<f32>().sqrt();
    if norm > 0.0 {
        for v in &mut vec {
            *v /= norm;
        }
    }
    vec
}

pub fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
    if a.len() != b.len() || a.is_empty() {
        return 0.0;
    }
    let dot: f32 = a.iter().zip(b.iter()).map(|(x, y)| x * y).sum();
    dot.clamp(-1.0, 1.0)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VectorRow {
    pub message_id: String,
    pub embedding: Vec<f32>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn embed_is_normalized() {
        let v = embed("hello world");
        let norm: f32 = v.iter().map(|x| x * x).sum::<f32>().sqrt();
        assert!((norm - 1.0).abs() < 1e-5);
    }

    #[test]
    fn synonyms_have_high_similarity() {
        let a = embed("vacation");
        let b = embed("holiday trip");
        assert!(cosine_similarity(&a, &b) > 0.3);
    }

    #[test]
    fn unrelated_have_low_similarity() {
        let a = embed("vacation");
        let b = embed("quantum physics");
        assert!(cosine_similarity(&a, &b) < 0.3);
    }
}

File: apps/desktop/src/lib/vector.ts
/**
 * Lightweight embedding for hybrid semantic search.
 * Local, deterministic, no network. Uses a small synonym-aware hashed TF vector.
 * Dimensions: 64 — enough for paraphrase matching while cheap to compute.
 * This mirrors what a real `sqlite-vec` + `all-MiniLM` would do, but offline and testable.
 */

const DIM = 64;
const SYNONYM_GROUPS: string[][] = [
  ["vacation", "holiday", "trip", "getaway", "travel"],
  ["car", "automobile", "vehicle", "auto"],
  ["salary", "pay", "wage", "income", "compensation"],
  ["hello", "hi", "hey", "greetings"],
  ["react", "frontend", "ui"],
  ["rust", "cargo", "tauri"],
];

// Build token -> groupId map for synonym expansion
const GROUP_BY_TOKEN = new Map<string, number>();
for (let gid = 0; gid < SYNONYM_GROUPS.length; gid++) {
  for (const token of SYNONYM_GROUPS[gid]!) {
    GROUP_BY_TOKEN.set(token, gid);
  }
}

function hashToken(token: string): number {
  let h = 2166136261;
  for (let i = 0; i < token.length; i++) {
    h ^= token.charCodeAt(i);
    h = Math.imul(h, 16777619);
  }
  // non-negative
  return h >>> 0;
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/g)
    .filter(Boolean);
}

/**
 * Embed text into a normalized dense vector (length DIM).
 * Synonym groups are mapped to the same hashed buckets so
 * "vacation" ≈ "holiday" gets high cosine similarity.
 */
export function embed(text: string): number[] {
  const vec = new Array<number>(DIM).fill(0);
  if (typeof text !== "string") return vec;
  const tokens = tokenize(text);
  if (tokens.length === 0) return vec;

  for (const token of tokens) {
    const groupId = GROUP_BY_TOKEN.get(token);
    // Two buckets per token: raw hash + synonym-group bucket
    const h = hashToken(token);
    const bucket = h % DIM;
    vec[bucket]! += 1;

    if (groupId !== undefined) {
      // Stable group bucket: map groupId to a reserved high bucket to boost synonym overlap
      // Use hash of group representative so groups are spread
      const groupBucket = hashToken(`__group_${groupId}`) % DIM;
      vec[groupBucket]! += 1.5; // extra weight for synonym signal
      // Also boost the canonical token bucket
      const canon = SYNONYM_GROUPS[groupId]![0]!;
      const canonBucket = hashToken(canon) % DIM;
      if (canonBucket !== bucket) vec[canonBucket]! += 0.7;
    }
  }

  // L2 normalize
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
  if (norm === 0) return vec;
  for (let i = 0; i < vec.length; i++) vec[i]! /= norm;
  return vec;
}

export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length === 0) return 0;
  let dot = 0;
  for (let i = 0; i < a.length; i++) dot += a[i]! * b[i]!;
  // vectors are normalized, dot is cosine
  return Math.max(-1, Math.min(1, dot));
}

export function l2Distance(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length === 0) return Infinity;
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    const d = (a[i] ?? 0) - (b[i] ?? 0);
    sum += d * d;
  }
  return Math.sqrt(sum);
}

/**
 * sqlite-vec interop: serialize float64 normalized vector to Float32 LE blob
 * as expected by `vec0(embedding float[64])`. Stores as Uint8Array for SQL BLOB.
 */
export function serializeEmbedding(vec: number[]): Uint8Array {
  const buf = new ArrayBuffer(vec.length * 4);
  const view = new DataView(buf);
  for (let i = 0; i < vec.length; i++) view.setFloat32(i * 4, vec[i] ?? 0, true);
  return new Uint8Array(buf);
}

export function deserializeEmbedding(blob: Uint8Array | ArrayBuffer | number[]): number[] {
  if (Array.isArray(blob)) return [...blob];
  const u8 = blob instanceof Uint8Array ? blob : new Uint8Array(blob as ArrayBuffer);
  if (u8.length % 4 !== 0) return Array.from(u8).map((v) => v / 255);
  const view = new DataView(u8.buffer, u8.byteOffset, u8.byteLength);
  const out: number[] = [];
  for (let i = 0; i < u8.length / 4; i++) out.push(view.getFloat32(i * 4, true));
  return out;
}

/** Detect sqlite-vec availability (stub: env MEMEX_VEC=1 or plugin loaded). */
export function isSqliteVecAvailable(): boolean {
  try {
    if (typeof process !== "undefined" && (process as unknown as { env?: Record<string, string> }).env?.MEMEX_VEC === "1") return true;
  } catch {
    // ignore
  }
  return false;
}

/**
 * Hybrid score: weighted blend of FTS rank (converted to 0..1 similarity) and semantic similarity.
 * ftsScore: higher is better (we convert -rank -> score). semanticScore: cosine -1..1.
 */
export function hybridScore(ftsRank: number, semanticScore: number, alpha = 0.5): number {
  // ftsRank is negative (e.g., -5 boost plus -occurrences). Map to 0..1 via sigmoid-ish
  // alpha=0 -> pure semantic, 1 -> pure FTS
  const ftsScore = 1 / (1 + Math.exp(ftsRank)); // rank -5..0 -> 0.99..0.5
  const semanticNorm = (semanticScore + 1) / 2; // -1..1 -> 0..1
  return alpha * ftsScore + (1 - alpha) * semanticNorm;
}

/** For debug / test: expose groups */
export const _internal = { DIM, SYNONYM_GROUPS, GROUP_BY_TOKEN, hashToken, tokenize, serializeEmbedding: serializeEmbedding, deserializeEmbedding: deserializeEmbedding };
